primary/src/tests/header_waiter_tests.rs

```diff
@@ -0,0 +1,134 @@
+        .send(WaiterMessage::SyncBatches(
+            missing_digests_map,
+            header.clone(),
+        ))
+        .await
+        .unwrap();
+
+    // AND spin up a worker node that primary owns
+    let worker_address = committee
+        .worker(&name, &worker_id)
+        .unwrap()
+        .primary_to_worker;
+
+    let handle = worker_listener::<PrimaryWorkerMessage<Ed25519PublicKey>>(1, worker_address);
+
+    // THEN
+    if let Ok(Ok(mut result)) = timeout(Duration::from_millis(4_000), handle).await {
+        match result.remove(0) {
+            PrimaryWorkerMessage::Synchronize(missing, _) => {
+                assert_eq!(
+                    missing_digests, missing,
+                    "Expected missing digests don't match"
+                );
+
+                // now simulate the write of the batch to the payload store
+                payload_store
+                    .write_all(missing_digests.into_iter().map(|e| ((e, worker_id), 1)))
+                    .await
+                    .unwrap();
+            }
+            _ => panic!("Unexpected message received!"),
+        }
+
+        // now get the output as expected
+        let header_result = rx_core.recv().await.unwrap();
+        assert_eq!(header.digest(), header_result.digest());
+    } else {
+        panic!("Messages not received by worker");
+    }
+}
+
+pub fn worker_listener<T>(
+    num_of_expected_responses: i32,
+    address: multiaddr::Multiaddr,
+) -> JoinHandle<Vec<T>>
+where
+    T: Send + DeserializeOwned + 'static,
+{
+    tokio::spawn(async move {
+        let mut recv = PrimaryToWorkerMockServer::spawn(address);
+        let mut responses = Vec::new();
+
+        loop {
+            match timeout(Duration::from_secs(1), recv.recv()).await {
+                Err(_) => {
+                    // timeout happened - just return whatever has already
+                    return responses;
+                }
+                Ok(Some(message)) => {
+                    match deserialize::<'_, T>(&message.payload) {
+                        Ok(msg) => {
+                            responses.push(msg);
+
+                            // if -1 is given, then we don't count the number of messages
+                            // but we just rely to receive as many as possible until timeout
+                            // happens when waiting for requests.
+                            if num_of_expected_responses != -1
+                                && responses.len() as i32 == num_of_expected_responses
+                            {
+                                return responses;
+                            }
+                        }
+                        Err(err) => {
+                            panic!("Error occurred {err}");
+                        }
+                    }
+                }
+                //  sender closed
+                _ => panic!("Failed to receive network message"),
+            }
+        }
+    })
+}
```

Comment:
Could you move the `worker_listener` you have in primary/src/block_synchronizer/tests/block_synchronizer_tests.rs to primary/src/tests/common.rs, and reuse it here and in the block synchronizer?

Reply:
Done in https://github.com/MystenLabs/narwhal/pull/297/commits/f16b900c3260ce2d537ddd92e37d82ca783248af

Reply:
Thanks a lot for https://github.com/MystenLabs/narwhal/commit/01acea86c6f5a15abacf549e2905da9cb8bf392f